### PR TITLE
feat: add lambda deployer additional user config function parameters

### PR DESCRIPTION
**Pull request by jsetton**
*Description of changes:*

This change adds the ability to customize some additional Lambda function config parameters (function name, description, memory size, timeout and environmental variables) through the ask resources user config, when using the Lambda deployer. This is useful if a user only needs to deploy a Lambda function part of the skill stack without the need to use the CFN deployer just to set this customization.

Some minor code refactoring and cleanup is included mostly moving the generate AWS resource name functions out of the clients library into the relevant builtins instead.

```
{
  "askcliResourcesVersion": "2020-03-31",
  "profiles": {
    "default": {
      "skillMetadata": {
        "src": "./skill-package"
      },
      "code": {
        "default": {
          "src": "./lambda"
        }
      },
      "skillInfrastructure": {
        "type": "@ask-cli/lambda-deployer",
        "userConfig": {
          "runtime": "nodejs14.x",
          "handler": "index.handler",
          "lambda": {
            "functionName": "<functionName>",
            "description": "<description>",
            "memorySize": 128,
            "timeout": 10,
            "environmentVariables": {
              "ENV_1": "<value1>",
              "ENV_2": "<value2>"
            }
          }
        }
      }
    }
  }
}
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by jsetton** [comment]
@RonWang any chance for this PR to be reviewed?

**Review by tydonelson — APPROVED** [review]

File: lib/builtins/deploy-delegates/lambda-deployer/helper.js [hunk 1]
```diff
@@ -232,13 +248,48 @@ function _addEventPermissions(lambdaClient, skillId, functionArn, profile, callb
     });
 }
 
+function _getLambdaCustomValue(alexaRegion, userConfig, parameter) {
+    return R.path(['regionalOverrides', alexaRegion, 'lambda', parameter], userConfig)
+    || R.path(['lambda', parameter], userConfig);
+}
+
+function _getLambdaFunctionName(alexaRegion, userConfig, skillName, profile) {
```
Comment (CamdenFoucht) [inline lib/builtins/deploy-delegates/lambda-deployer/helper.js:256]: nit: Would be nice to have a quick jsdoc comments above these new methods that describe it, just something like:

```
/*
 * Loads the function name from config or generates a unique function name
 *
 */
```

File: lib/builtins/deploy-delegates/lambda-deployer/helper.js [hunk 2]
```diff
@@ -232,13 +248,48 @@ function _addEventPermissions(lambdaClient, skillId, functionArn, profile, callb
     });
 }
 
+function _getLambdaCustomValue(alexaRegion, userConfig, parameter) {
```
Comment (CamdenFoucht) [inline lib/builtins/deploy-delegates/lambda-deployer/helper.js:251]: nit: Would be great to have jsdoc comments

File: lib/builtins/deploy-delegates/lambda-deployer/helper.js [hunk 3]
```diff
@@ -267,42 +318,44 @@ function _updateLambdaFunction(reporter, lambdaClient, options, callback) {
 function _updateLambdaFunctionCode(reporter, lambdaClient, options, callback) {
     const { zipFilePath, deployState } = options;
     const zipFile = fs.readFileSync(zipFilePath);
-    const functionName = deployState.lambda.arn;
+    const functionArn = deployState.lambda.arn;
```
Comment (CamdenFoucht) [inline lib/builtins/deploy-delegates/lambda-deployer/helper.js:321]: Much better :D

File: lib/builtins/deploy-delegates/lambda-deployer/helper.js [hunk 4]
```diff
@@ -232,13 +248,48 @@ function _addEventPermissions(lambdaClient, skillId, functionArn, profile, callb
...
+    const memorySize = _getLambdaCustomValue(alexaRegion, userConfig, 'memorySize');
+    const timeout = _getLambdaCustomValue(alexaRegion, userConfig, 'timeout');
+    const environmentVariables = _getLambdaCustomValue(alexaRegion, userConfig, 'environmentVariables');
+    return {
+        runtime,
+        handler,
+        description,
+        // default values to account for Java runtime cold start
+        memorySize: Number.parseInt(memorySize, 10) || 512,
+        timeout: Number.parseInt(timeout, 10) || 15,
```
Comment (CamdenFoucht) [inline lib/builtins/deploy-delegates/lambda-deployer/helper.js:281]: nit: Should we put the memorySize and timeout in the constants file? (I know this is just moved from pre-existing code)

File: lib/clients/aws-client/lambda-client.js [hunk 5]
```diff
@@ -16,24 +15,25 @@ module.exports = class LambdaClient extends AbstractAwsClient {
...
             Runtime: runtime,
             Handler: handler,
-            // to account for Java runtime cold start
-            Timeout: 15,
-            MemorySize: 512,
-            FunctionName: this._generateFunctionName(skillName, profile, alexaRegion)
+            ...(description && { Description: description }),
+            MemorySize: memorySize,
+            Timeout: timeout,
+            ...(environmentVariables && { Environment: { Variables: environmentVariables } })
```
Comment (CamdenFoucht) [inline lib/clients/aws-client/lambda-client.js:36]: nit: I think keeping the optional/possibly undefined stuff would make this object a little bit cleaner

```
const params = {
   Code: {
                ZipFile: zipFile
    },
     FunctionName: functionName,
     Handler: handler,
     MemorySize: memorySize,
     Role: role,
     Runtime: runtime,
     Timeout: timeout,
     ...(description && { Description: description }),
     ...(environmentVariables && { Environment: { Variables: environmentVariables } 
})
```

File: test/unit/clients/aws-client/lambda-client-test.js [hunk 6]
```diff
@@ -66,8 +64,7 @@ describe('Clients test - lambda client test', () => {
             const TEST_CREATE_FUNCTION_ERR = 'create_function_error';
             sinon.stub(lambdaClient.client, 'createFunction').callsArgWith(1, TEST_CREATE_FUNCTION_ERR);
             // call
-            lambdaClient.createLambdaFunction(TEST_SKILL_NAME, TEST_AWS_PROFILE, TEST_ALEXA_REGION,
-                TEST_IAM_ROLE, TEST_ZIPFILD, TEST_RUNTIME, TEST_HANDLE, (err) => {
+            lambdaClient.createLambdaFunction(TEST_FUNCTION_NAME, TEST_FUNCTION_CONFIG, TEST_IAM_ROLE, TEST_ZIPFILE, (err) => {
```
Comment (CamdenFoucht) [inline test/unit/clients/aws-client/lambda-client-test.js:67]: Thanks for updating the unit tests!

File: lib/builtins/deploy-delegates/lambda-deployer/helper.js [hunk 7]
```diff
@@ -127,6 +128,20 @@ function _createIAMRole(reporter, iamClient, skillName, callback) {
     });
 }
 
+function _getIAMRoleName(skillName) {
```
Comment (CamdenFoucht) [inline lib/builtins/deploy-delegates/lambda-deployer/helper.js:131]: nit: jsdoc comment for the function 😄

File: docs/concepts/Deploy-Command.md [hunk 8]
```diff
@@ -115,11 +115,16 @@ This deployer is managing Lambda services by using the default settings to creat
...
 Supported parameters in `userConfig`:
-* awsRegion
-* runtime: Please see [AWS Lambda runtimes](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html) for more details.
-* handler: The entry point for the Lambda function in your code.
-* sourceLambda: You can set the `userConfig.sourceLambda.arn` to specify an existing Lambda ARN to reuse. This will be the target Lambda function when you deploy.
-* regionalOverrides: You can set `regionalOverrides.{region}.{anyPropertiesAbove}` to override the parameters for a certain region.
+* `awsRegion`
+* `runtime`: Please see [AWS Lambda runtimes](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html) for more details.
+* `handler`: The entry point for the Lambda function in your code.
+* `lambda.functionName`: The custom name for the Lambda function, otherwise ask cli will generate a unique name based on the skill, profile and region names.
```
Comment (CamdenFoucht) [inline docs/concepts/Deploy-Command.md:121]: Great stuff!

**Review by CamdenFoucht — APPROVED** [review]
Thanks again for the PR, letting people have this config for their lambdas should definitely be supported!

Everything looked great! I left some nits mostly on function comments, but none of them are required.

**Pull request merged**